### DISCUSSION TITLE
Using manufacturer lens correction data?

### DISCUSSION THREAD

**Discussion by uchrisu** [lensfun/lensfun#1606] [Ideas]
As we all know, creating lens calibrations is a time consuming task and the potential for mistakes is large.

Nowadays, manufactures often include correction data into their lenses. What is your opinion on using them? 
Probably they are not directly compatible to our vignetting-/ca-/distortion models, but we can convert them by best fit. 

Would you agree that this is an alternative, at least for many modern lenses?

I will try this for my e-mount lenses but before I spend to much effort in it, I would know if you agree to use such data.

**Comment by Sturmflut** [comment]
How is the data stored and represented? EXIF tags?

**Comment by uchrisu** [comment]
For Sony there is the data in the Exif tags.

The idea (after verifying it) is to make some pictures with different focus/aperture/focal length, put them in a directory, run a script an get the corresponding lensfun.xml

**Comment by Sturmflut** [comment]
Interesting. If a reliable conversion/fitting algorithm could be found, lensfun could maybe even do this at runtime, without offline conversion...

Do you know which EXIF tags are being used? I don't have any Sony gear and have never looked closely at their EXIF tags. Is this present in all lenses?

**Comment by uchrisu** [comment]
E-mount lenses have these parameters: 
Vignetting Corr Params         
Chromatic Aberration Corr Params
Distortion Corr Params    

I can only check for sony, but probably there are equivalents for other manufacturers. 

These parameters are per image. So they are calculated for a specific scene (focal length, aperture, ...) and at the moment I see no way to get the underlying parameters that calculate these parameters. So our way of fitting and interpolating will be the way to go. 

As far as I know, many lenses have these parameters set, but not all.

**Comment by sarunasb** [comment]
Yes, lens manufacturers are incorporating correction data in RAW files 
for quite some time by now. Anyone remembers Bibble Pro? The only more 
or less usable Linux/open source RAW development software 10–15 years 
ago? It used to maintain its own lens corrections database, modest in 
coverage, and that's when first (to my memory) the fact that µ4/3 
(Olympus, Panasonic) have some correction data included in RAW became 
known. The problem was of course that manufacturers only cooperated with 
Adobe and alike as to how to use that data. It was when proliferation of 
“digital” lenses was starting — instead of designing insanely complex, 
optically perfect lenses, why not correct known deficiencies of those 
lenses in post-processing?

So in many cases the data is there, but, as far as I know, not only 
there is no known standard, but even proprietary “standards” are not 
published. Correction values included in RAW then will have to be 
“deciphered” per manufacturer. That can also change any time, per lens 
or per camera firmware version, when manufacturer seems fit to, and of 
course there will be no announcement about that. I wish I'm wrong and 
there is actually a standard for corrections data, to which lens 
manufacturers adhere to. If not now, then maybe in future.

In the end, this might still be very useful and worth the effort on a 
case-by-case basis, if it is known how to extract correction data for 
certain class of lenses (combination of manufacturer/camera FW/lens), 
and then to convert it into model supported by Lensfun. I'm a bit 
skeptical about doing this at runtime as correction logic would have to 
be maintained in Lensfun codebase per manufacturer, at least. Not that 
it's impossible.


-- 
Šarūnas Burdulis
math.dartmouth.edu/~sarunas

**Reply by uchrisu** [reply]
It fully agree to what @sarunasb  wrote.

Thus, I will try to create standard lensfun profiles from this data in a first step, beginning with Sony and maybe later others will follow.

**Reply by tuxfanx** [reply]
...I found this  #1304 
maybe it is interesting for you

**Reply by uchrisu** [reply]
Thanks for this hint. I found some familiar points but I'm not sure if everything in #1304 is correct in detail but I didn't go through it in detail yet.

**Comment by uchrisu** [comment]
Okay, since the coefficients in the exif use a different scaling I scaled them to be compatible with lensfun, i.e., the normalizes radius 1 stays unscaled. Just to remember: The normalized radius r=1 is half the image height. 
Afterwards I fitted a LF_DIST_MODEL_PTLENS  to this data. Furthermore I compared the distortion correction to coefficients I found with a conventional "line-drawing" on the sample image. 

Here is the result:
![lensfun_extract_exif_example1](https://user-images.githubusercontent.com/9560869/153773067-1ceecf1b-2620-441d-bf15-f890db9907db.png)

We see that we can fit the data quite good with the LF_DIST_MODEL_PTLENS.
However, we see deviations to the conventional method for smaller r. I think the conventional fitting with two lines is the weak candidate here. Up to approx. r=1/3 we have absolutely no data with the two-line method and in the range slightly above the information we have from the inner line is rather loose. 
So probably we should trust the exif data here. 

The results (coefficients applied in darktable) are both acceptable. They mostly differ in the automatic overall scaling:
Conventional:
![conventional](https://user-images.githubusercontent.com/9560869/153773363-7dc601d6-773f-497e-b7ed-c9f0e47b4c14.jpg)

With fitted exif-data:
![fitted coeffs](https://user-images.githubusercontent.com/9560869/153773396-8bd2aa5f-c5f9-4b3c-854d-daee16fdd7d3.jpg)

**Reply by sarunasb** [reply]
Looks well to me, i.e. images' difference as far as distortion is negligable.

**Comment by uchrisu** [comment]
Vignetting:

When I looked at the data I had an intuitive guess how it could work and this is the result:
![figure_vig1](https://user-images.githubusercontent.com/9560869/154133156-88193589-9076-4054-9b05-e108e6b6e4a4.png)

The conventional curve is taken from the lensfun database. The curves look similar but in the middle there are some differences. So I took a testshot and used the conventional and the new fitted model in darktable to correct the vignetting of the even lit image:

Without correction:
![vig_orig](https://user-images.githubusercontent.com/9560869/154133566-a2a950cb-f408-4f64-b5ed-5c57c7652b93.png)

From LF database:
![vig_conv](https://user-images.githubusercontent.com/9560869/154133633-b3df2418-378f-4950-8a4d-d4c20e8abd56.png)

Model fitted to in-camera data:
![vig_fit](https://user-images.githubusercontent.com/9560869/154133690-eab52db3-94c7-4736-b334-b5d6c8baff85.png)

So, the new model from the in-camera data does a better job here. Thus, I think we can also use it for vignetting.

Next to go: TCA

**Comment by uchrisu** [comment]
TCA:

I found some information here:
[https://discuss.pixls.us/t/sony-raw-chromatic-aberration-correction-model/21153/17](url)

I implemented my own version and got these results for example:
no correction:
![PEN08772_no_crop](https://user-images.githubusercontent.com/9560869/154557925-973b0dbf-7086-44db-9933-3210746ed48e.jpg)
with correction (extracted and fitted data): 
![PEN08772_crop](https://user-images.githubusercontent.com/9560869/154558026-0761ce74-682e-4862-8ed3-9734233453fc.jpg)

Another image with 300% crop:
no correction:
![PEN08761_no_300p](https://user-images.githubusercontent.com/9560869/154558096-2024fe2c-2d76-416f-961b-d829589cfbf0.jpg)
with correction:
![PEN08761_300p](https://user-images.githubusercontent.com/9560869/154558144-0934f402-de51-4f36-8151-e2f92c1ac4c3.jpg)

Seems to work.

Next goal: Write a script that extracts the data automatically and scans a whole directory for images and creates lensfun.xml file with the extracted and fitted data.

**Comment by uchrisu** [comment]
I created a little python script that does all the work.
Simply execute it in the directory where also the directories distortion, tca and vignetting are and wait for the results. 
The actual content of the images does not matter. Only the metadata has to be correct, i.e., take pictures at the right apertures/focal lengths/distances with arbitrary content. However, you can take meaningful content to check the results. 

[https://github.com/uchrisu/lf_fitexif](https://github.com/uchrisu/lf_fitexif)

Currently it only works for Sony E-Mount lenses and is tested only with full frame lenses. 
For my test lens the results coincide very well with the manually achieved coefficients.

**Reply by sarunasb** [reply]
Nice work!, @uchrisu 

Calibrators usually add a comment to `<calibration>` element: `<!-- Taken with ...>` to mention the camera, from which sample images where used. Would it make sense to add something similar from your script? That data was computed from Sony's EXIF or something. Up to you of course.

**Reply by uchrisu** [reply]
That is a good idea. I will add this very soon.
